pairing/bn254/point.go

```diff
@@ -0,0 +1,786 @@
+	// Clone is required as we change the point
+	p = p.Clone().(*pointG1)
+
+	n := p.ElementSize()
+	// Take a copy so that p is not written to, so calls to MarshalBinary
+	// are threadsafe.
+	pgtemp := *p.g
+	pgtemp.MakeAffine()
+	ret := make([]byte, p.MarshalSize())
+	if pgtemp.IsInfinity() {
+		return ret, nil
+	}
+	tmp := &gfP{}
+	montDecode(tmp, &pgtemp.x)
+	tmp.Marshal(ret)
+	montDecode(tmp, &pgtemp.y)
+	tmp.Marshal(ret[n:])
+	return ret, nil
+}
+
+func (p *pointG1) MarshalID() [8]byte {
+	return marshalPointID1
+}
+
+func (p *pointG1) MarshalTo(w io.Writer) (int, error) {
+	buf, err := p.MarshalBinary()
+	if err != nil {
+		return 0, err
+	}
+	return w.Write(buf)
+}
+
+func (p *pointG1) UnmarshalBinary(buf []byte) error {
+	n := p.ElementSize()
+	if len(buf) < p.MarshalSize() {
+		return errors.New("bn254.G1: not enough data")
+	}
+	if p.g == nil {
+		p.g = &curvePoint{}
+	} else {
+		p.g.x, p.g.y = gfP{0}, gfP{0}
+	}
+
+	p.g.x.Unmarshal(buf)
+	p.g.y.Unmarshal(buf[n:])
+	montEncode(&p.g.x, &p.g.x)
+	montEncode(&p.g.y, &p.g.y)
+
+	zero := gfP{0}
+	if p.g.x == zero && p.g.y == zero {
+		// This is the point at infinity
+		p.g.y = *newGFp(1)
+		p.g.z = gfP{0}
+		p.g.t = gfP{0}
+	} else {
+		p.g.z = *newGFp(1)
+		p.g.t = *newGFp(1)
+	}
+
+	if !p.g.IsOnCurve() {
+		return errors.New("bn254.G1: malformed point")
+	}
+
+	return nil
+}
+
+func (p *pointG1) UnmarshalFrom(r io.Reader) (int, error) {
+	buf := make([]byte, p.MarshalSize())
+	n, err := io.ReadFull(r, buf)
+	if err != nil {
+		return n, err
+	}
+	return n, p.UnmarshalBinary(buf)
+}
+
+func (p *pointG1) MarshalSize() int {
+	return 2 * p.ElementSize()
+}
+
+func (p *pointG1) ElementSize() int {
+	return 256 / 8
+}
+
+func (p *pointG1) String() string {
+	return "bn254.G1" + p.g.String()
+}
+
+func (p *pointG1) Hash(m []byte) kyber.Point {
+	return hashToPoint(p.dst, m)
+}
+
+func hashToPoint(domain, m []byte) kyber.Point {
+	e0, e1 := hashToField(domain, m)
+	p0 := newPointG1(domain).fromBigInt(mapToPoint(e0))
+	p1 := newPointG1(domain).fromBigInt(mapToPoint(e1))
+	p := p0.Add(p0, p1)
+	return p
+}
+
+func hashToField(domain, m []byte) (*big.Int, *big.Int) {
+	const u = 48
+	_msg := expandMsgXmd(domain, m, 2*u)
+	x := new(big.Int)
+	y := new(big.Int)
+	x.SetBytes(_msg[0:48]).Mod(x, p)
+	y.SetBytes(_msg[48:96]).Mod(y, p)
+	return x, y
+}
+
+// `mapToPoint` implements a specialised SW mapping for BN curves from the paper
+//
+//	 Fouque, P.-A. and M. Tibouchi, "Indifferentiable Hashing to Barreto--Naehrig Curves",
+//		In Progress in Cryptology -
+//		LATINCRYPT 2012, pages 1-17,
+//		DOI 10.1007/978-3-642-33481-8_1, 2012,
+//		<https://doi.org/10.1007/978-3-642-33481-8_1>.
+//
+// Ref implementations:
+//
+//	https://github.com/herumi/mcl/blob/5f4449efd08388009f9abce06c44fc26730193e7/include/mcl/bn.hpp#L343
+//	https://github.com/thehubbleproject/hubble-contracts/blob/f1c13fe4e1a0dc9ab1f150895de7c0e654ee46b0/contracts/libs/BLS.sol#L139
+func mapToPoint(x *big.Int) (*big.Int, *big.Int) {
```

Comment:
How different is this from the simplified SW mapping https://datatracker.ietf.org/doc/html/rfc9380#section-6.6.2 ?
Seems Fouque-Tibouchi is a making the SW construction explicit in the case of BN curves, and according to the [SSWU paper for BLS](https://eprint.iacr.org/2019/403.pdf), it seems that it's really the same as the SSWU but for BN curves:

> Fouque and Tibouchi [FT12] also analyze the original construction of Shallue and van de Woestijne
specifically for the case of Barreto-Naehrig curves [BN06]; in Section 3, we give a very similar construction tailored to the BLS12-381 curve

So I guess it's the same but for BN. And so it counts as SSWU ✅

Reply:
Fouque-Tibouchi references the [Shallue-van de Woestijne paper](https://link.springer.com/chapter/10.1007/11792086_36) and not the [Ulas paper](https://www.impan.pl/pl/wydawnictwa/czasopisma-i-serie-wydawnicze/bulletin-polish-acad-sci-math/all/55/2/85475/rational-points-on-certain-hyperelliptic-curves-over-finite-fields) so I think that this is actually the "SVDW" mapping described in [6.6.1](https://datatracker.ietf.org/doc/html/rfc9380#svdw) rather than SSWU.